Tranpiler for town daily tick

thomas-soutif opened the pull request:
handle the case of the foodStock of a fief
I should modify it because the food stack value is not exactly the one we want, but Im' in good progress

## PR Checklist
<!-- Insert "x" inside square brackets to check item. -->

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR. -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:


## What is the current behaviour?
<!-- Please describe the current behaviour that you are modifying, or link to a relevant issue. -->


## What is the new behaviour?
<!-- Insert issue id after hashtag. -->
Resolves #757 
<!-- Describe functionality added. Add images or videos to show how it works if applies -->


## Other information

thomas-soutif committed:
Tranpiler for town daily tick

handle the case of the foodStock of a fief
I should modify it because the food stack value is not exactly the one we want, but Im' in good progress

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,51 @@
+    [HarmonyTranspiler]
+    private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instrs)
+    {
+        List<CodeInstruction> instructions = instrs.ToList();
+        int setterLastIdx = instructions.FindLastIndex(i => i.opcode == Call.opcode && i.operand as MethodInfo == foodStocksSetter);
+        if (setterLastIdx == -1) return instrs;
+
+        instructions.InsertRange(setterLastIdx, new CodeInstruction[]
+        {
+            new CodeInstruction(OpCodes.Ldarg_0),
+            new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(TownDailyTickPatch), "LogFoodStock"))
+        });
+        
+        return instructions;
```
Currently this will only patch the last FoodStocks set

This also does not block setting on the client

Below will replace the call with your method

The stack will also contain the new food stocks value so you will have to set it yourself for more control

```suggestion
        foreach (var instriction in instrs)
{
    if (i.opcode == Call.opcode && i.operand as MethodInfo == foodStocksSetter)
    {
        yield return new CodeInstruction(OpCodes.Ldarg_0);
        yield return new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(TownDailyTickPatch), "LogFoodStock"));
        continue;
    }

    yield return instriction
}
```

thomas-soutif replied:
When I remplace with this code, I have this crash :

![image](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/assets/23268707/14cb0acf-1232-4d38-95f7-97ff80db47a5)

thomas-soutif replied:
The crash seem to come from the yield return

thomas-soutif replied:
So I was able to make it work without crash by removing the new code insruction(OpCodes.Ldarg0)
the crash occurs because we add two new code instruction for one instruction replace
but now I have this as a result : 
![image](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/assets/23268707/4c773c5c-5525-4969-8725-81ee47f866a2)
the fief instance return here for exemple is a list of "village", maybe because 3 villages have the same food stock in same time ?

thomas-soutif replied:
Finally make it works !
```
    foreach (var instruction in instrs)
    {
        if (instruction.opcode == Call.opcode && instruction.operand as MethodInfo == foodStocksSetter)
        {
	    yield return new CodeInstruction(OpCodes.Call, AccessTools.Method(typeof(TownDailyTickPatch), "LogFoodStock"));
            //yield return new CodeInstruction(OpCodes.Ldarg_0);
            continue;
        }
        yield return instruction;
    }
    
}

public static void LogFoodStock(Fief fief, float value)
```
so I remove the Ldarg_0 and change the order of the parameter that take LogFoodStock, the problem was there

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,51 @@
+        
+    }
+
+    public static void LogFoodStock(Fief fief)
```
Update to match new patch

```suggestion
    public static void LogFoodStock(float newValue, Fief fief)
```

thomas-soutif replied:
Because we get from the stack the Ldarg_0, that represent the "fief" instance , putting a float newValue make the program crash
we should found the value return by the previous stack (get_foodStack) in the dailyTick() method
![image](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/assets/23268707/00af9992-a105-4d19-bd43-76c8fee7d1d9)

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,51 @@
+    public static void LogFoodStock(Fief fief)
+    {
+        // If it's the client, return
+        if (!ModInformation.IsServer) return;
```
No '!' is less confusing at a glance

```suggestion
        if (ModInformation.IsClient) return;
```

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,51 @@
+        // If it's the client, return
+        if (!ModInformation.IsServer) return;
+        
+        var message = new FiefFoodStockChanged(fief.StringId, fief.FoodStocks);
```
```suggestion
        fief.FoodStocks = newValue;
        var message = new FiefFoodStockChanged(fief.StringId, fief.FoodStocks);
```

thomas-soutif committed:
fix the dailytick transpiler : food sync works now

The food stock of a fief was not synchronizing well, with the transpiler that replace the original setter by a custom method, we can now handle our self the setter of the FoodStock

thomas-soutif committed:
ModInformation.IsClient

 better lisibility

thomas-soutif committed:
clean some comments

thomas-soutif commented:
Bug fixed, ready for review

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,50 @@
+        
+        foreach (var instruction in instrs)
+        {
+            if (instruction.opcode == Call.opcode && instruction.operand as MethodInfo == foodStocksSetter)
```
```suggestion
            if (instruction.opcode == Call.opcode && instruction.operand as MethodInfo == AccessTools.PropertySetter(typeof(Fief), "FoodStocks"))
```

garrettluskey commented on source/GameInterface/Services/Towns/Patches/DailyTickTownPatch.cs:
```diff
@@ -0,0 +1,50 @@
+[HarmonyPatch(typeof(Town), "DailyTick")]
+public static class TownDailyTickPatch
+{
+    // Get the PropertyInfo for the FoodStocks property
+    private static PropertyInfo foodStocksProperty = typeof(Fief).GetProperty("FoodStocks");
+
+    // Get the setter method for the FoodStocks property
+    private static MethodInfo foodStocksSetter = foodStocksProperty.GetSetMethod(true);
```
This data is stored the entire lifetime of the program, if they are not used outside the transpiler use the "=>" operator or move it into the transpiler like suggested
```suggestion
```

thomas-soutif committed:
Use AccessTools Property setter

Co-authored-by: Garrett Luskey <garrett.luskey@gmail.com>

thomas-soutif committed:
removed static Property Info

Co-authored-by: Garrett Luskey <garrett.luskey@gmail.com>
